### PR TITLE
Added ImageOps contain()

### LINKED ISSUE #5415: Function similar to pad but returns resized image without padding
I'd sure like a function that looks like ImageOps.pad
https://pillow.readthedocs.io/en/stable/_modules/PIL/ImageOps.html#pad

but just returns the image after resizing without adding the padding.

What would be a good name for such a function? Where would it best be placed?

ImageOps.fit is already taken.

EDIT: Looking at #3364 `contain` seems like a decent candidate. Seems a refactor would be required. @radarhere thoughts?

### PR DESCRIPTION
**Pull request by radarhere**
Resolves #5415

Adds a `contain()` method to ImageOps. I would describe it as the inverse of `fit()`.

Running this code for each method in turn for comparison -

```python
from PIL import Image, ImageOps
im = Image.open("Tests/images/hopper.png")
ImageOps.fit(im, (200, 300))
```

|       | Fit                                                                                                        | Pad                                                                                                        | Contain                                                                                                        |
| ----- | ---------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------- |
| Size  |200 x 300                                                                                                   | 200 x 300                                                                                                  | 200 x 200                                                                                                      |
| Image |fit|pad|contain|

### CODE REVIEW

**Comment by radarhere** [comment]
3.10-dev failures seem to be due to https://github.com/pytest-dev/pytest/issues/8539, which has been resolved, but is not available in a pytest release yet.

**Comment by axu2** [comment]
Looks great, exactly what I was looking for.

**Comment by axu2** [comment]
OK, I actually copy pasted the code and the contain method seems to change the aspect ratio.

**Comment by axu2** [comment]
Moreover, I think the same problem exists with the pad function as well.

I am passing images that are skinnier than the requested `size`.

**Comment by axu2** [comment]
OK, I actually tried to run the pad function myself and I got some weird issues.

`im1` is a file thats 2037 wide 3056 tall. When I run this I get:

```
>>> ImageOps.pad(im1, (824,1200)).show() 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "...\venv\lib\site-packages\PIL\ImageOps.py", line 262, in pad
    image = image.resize((new_width, size[1]), resample=method)
  File "...\venv\lib\site-packages\PIL\Image.py", line 1888, in resize
    return self._new(self.im.resize(size, resample, box))
ValueError: height and width must be > 0
```

**Comment by axu2** [comment]
Sorry, I left a bunch of comments that I quickly deleted, since I had made an error.

File: src/PIL/ImageOps.py [hunk 1]
```diff
@@ -236,6 +236,34 @@ def colorize(image, black, white, mid=None, blackpoint=0, whitepoint=255, midpoi
     return _lut(image, red + green + blue)
 
 
+def contain(image, size, method=Image.BICUBIC):
+    """
+    Returns a sized version of the image, expanded to fill the requested aspect ratio
+    and size.
+
+    :param image: The image to size and crop.
```
Comment (hugovk) [inline src/PIL/ImageOps.py:244]:
```suggestion
    :param image: The image to resize and crop.
```

File: src/PIL/ImageOps.py [hunk 2]
```diff
@@ -236,6 +236,34 @@ def colorize(image, black, white, mid=None, blackpoint=0, whitepoint=255, midpoi
...
+    :param size: The requested output size in pixels, given as a
+                 (width, height) tuple.
+    :param method: What resampling method to use. Default is
```
Comment (hugovk) [inline src/PIL/ImageOps.py:247]:
```suggestion
    :param method: Resampling method to use. Default is
```

File: src/PIL/ImageOps.py [hunk 3]
```diff
@@ -236,6 +236,34 @@ def colorize(image, black, white, mid=None, blackpoint=0, whitepoint=255, midpoi
     return _lut(image, red + green + blue)
 
 
+def contain(image, size, method=Image.BICUBIC):
+    """
+    Returns a sized version of the image, expanded to fill the requested aspect ratio
+    and size.
```
Comment (hugovk) [inline src/PIL/ImageOps.py:242]: Maybe something like this is clearer?
```suggestion
    Returns a resized version of the image, expanded to fill the requested size by cropping, and with no padding.
```

File: Tests/test_imageops.py [hunk 4]
```diff
@@ -99,6 +102,21 @@ def test_fit_same_ratio():
         assert new_im.size == (1000, 755)
 
 
+def test_contain():
```
Comment (hugovk) [inline Tests/test_imageops.py:105]: I think these three cases can be refactored using `pytest.mark.parametrize`?

**Review by hugovk — COMMENTED** [review]
Let's also include this in the docs (`autofunction` in `docs/reference/ImageOps.rst`?) and release notes.

File: Tests/test_imageops.py (hunk 4 above)
Comment (radarhere) [reply Tests/test_imageops.py]: I've pushed a commit for this

File: src/PIL/ImageOps.py (hunk 3 above)
Comment (radarhere) [reply src/PIL/ImageOps.py]: There isn't any cropping in this method.
I've pushed a commit with
> Returns a resized version of the image, set to the maximum width and height within the requested size, while maintaining the original aspect ratio.

Feel free to suggest further changes.

Comment (hugovk) [reply src/PIL/ImageOps.py]: Sounds good, thanks! (I forgot hopper.png is square when looking at the examples.)

File: docs/releasenotes/8.3.0.rst [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+ImageOps.contain
+^^^^^^^^^^^^^^^^
+
+Returns a resized version of the image, set to the maximum width and height within
+``size``, while maintaining the original aspect ratio.
+
+To compare it to other ImageOps methods:
+- :py:meth:`~PIL.ImageOps.fit` expands an image until is fills ``size``, cropping the
+parts of the image that do not fit.
```
Comment (hugovk) [inline docs/releasenotes/8.3.0.rst:46]:
```suggestion
To compare it to other ImageOps methods:

- :py:meth:`~PIL.ImageOps.fit` expands an image until is fills ``size``, cropping the
parts of the image that do not fit.
```

Is a newline needed to activate bullet points in rst?

image

https://docutils.sourceforge.io/docs/user/rst/quickref.html#bullet-lists

Comment (radarhere) [reply docs/releasenotes/8.3.0.rst]: Yep, I guess newline, but also some indentation. Should be right now.

**Comment by hugovk** [comment]
Thank you!

**Comment by axu2** [comment]
🥳

**Comment by axu2** [comment]
Oh one last comment, just realized this function is similar to Image.thumbnail except it will upscale the image, if needed.

Also, for upscaling low resolution text, PIL.NEAREST might be a better default choice.
https://en.wikipedia.org/wiki/Comparison_gallery_of_image_scaling_algorithms

**Pull request merged**